### PR TITLE
Node data dump 2

### PR DESCRIPTION
**Pull request by LauraStelzner-TRI**
PR part 2 of node_data_dump
Added ability to dump the node data of each node within a tree.  Added prefix ability to introspection in case you want to focus on one branch of the tree instead of the entire tree.

### CODE REVIEW

File: src/task_behavior_ros/introspection.py [hunk 1]
```diff
@@ -146,6 +153,60 @@ def publish_status(self):
...
+        if status.status is not NodeStatus.PENDING:
+            ndd.key.append('status')
+            ndd.value.append(str(status))
+        return ndd
+
+    def _create_tree_data_dump_msg(self, node, msg=TreeDataDump()):
+        """ Create the status message
+            @param node [Node] The node to get the status from (recursive)
+             @param msg [TreeDataDump] The message created so far
+            @returns [TreeDataDump] The status message
```
Comment (tinawright) [inline src/task_behavior_ros/introspection.py]: nit: inconsistent formatting

File: src/task_behavior_ros/introspection.py [hunk 2]
```diff
@@ -146,6 +153,60 @@ def publish_status(self):
...
+        node_msg.id = str(node._id)
+        node_msg.name = node._name
+
+        data_msg = self._create_node_data_dump(node)
+        msg.data.append(data_msg)
+
+        if Behavior in type(node).__bases__:
+            node_msg.type = TreeNode.BEHAVIOR
+            for child in node._children:
+                node_msg.children.append(str(child._id))
```
Comment (tinawright) [inline src/task_behavior_ros/introspection.py]: you can also create this list all at once node_msg.children = [str(child._id) for child in node._children]
it's usually more efficient than looping https://wiki.python.org/moin/PythonSpeed/PerformanceTips#Loops

File: src/task_behavior_ros/introspection.py [hunk 3]
```diff
@@ -146,6 +153,60 @@ def publish_status(self):
...
+            msg.node.append(node_msg)
+            for child in node._children:
+                msg = self._create_tree_data_dump_msg(child, msg)
```
Comment (tinawright) [inline src/task_behavior_ros/introspection.py]: you're resetting and reusing msg for each child function call - is that intentional?

**Comment by LauraStelzner-TRI** [comment]
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/task_behavior_ros/introspection.py, line 185 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAgR-8sdeA99qFXw2U-r1-185:-KjAoYeVUo6p3oty16ff:b-896fix) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/src/task_behavior_ros/introspection.py#L185)):*
<details><summary><i>Previously, tinawright (Tina Wright (TRI)) wrote…</i></summary><blockquote>

nit: inconsistent formatting
</blockquote></details>

Done.

---

*[src/task_behavior_ros/introspection.py, line 197 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAgR-8sdeA99qFXw2U-r1-197:-KjAouXgul_u5ua9Ghnp:b-896fix) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/src/task_behavior_ros/introspection.py#L197)):*
<details><summary><i>Previously, tinawright (Tina Wright (TRI)) wrote…</i></summary><blockquote>

you can also create this list all at once node_msg.children = [str(child._id) for child in node._children]
it's usually more efficient than looping https://wiki.python.org/moin/PythonSpeed/PerformanceTips#Loops
</blockquote></details>

Done.

---

*[src/task_behavior_ros/introspection.py, line 200 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAgR-8sdeA99qFXw2U-r1-200:-KjAowryIT1NYt0_8lWF:b-datyfk) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/src/task_behavior_ros/introspection.py#L200)):*
<details><summary><i>Previously, tinawright (Tina Wright (TRI)) wrote…</i></summary><blockquote>

you're resetting and reusing msg for each child function call - is that intentional?
</blockquote></details>

I C&P from the other functions as a starting point, my understanding is it was designed that way because its a recursive function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10)*

**Comment by athackst** [comment]
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[task_behavior_ros.rosinstall, line 7 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAmaNZg0TeAjl2q6Jw:-KjAmaN_Pd8Yk7ZFOmQ_:b-l35sj5) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/task_behavior_ros.rosinstall#L7)):*
> ```
>     local-name: task_behavior_msgs
>     uri: https://github.com/ToyotaResearchInstitute/task_behavior_msgs.git
>     version: node_data_dump
> ```

this will need to be removed

---

*[task_behavior_ros.rosinstall, line 11 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAmeGsZ7wXCG33fhxX:-KjAmeGsZ7wXCG33fhxY:b-fqtemz) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/task_behavior_ros.rosinstall#L11)):*
> ```
>     local-name: task_behavior_ros
>     uri: https://github.com/ToyotaResearchInstitute/task_behavior_ros.git
>     version: node_data_dump
> ```

this will need to be removed (also this one shouldn't be necessary as CI should pull/build the branch you are on)

---

*[src/task_behavior_ros/introspection.py, line 35 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAnXQk7N5rOPPoJk_G:-KjAnXQluOSENTm1D2hA:b-w4bdwv) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/src/task_behavior_ros/introspection.py#L35)):*
> ```Python
>     """
> 
>     def __init__(self, node, prefix="tree"):
> ```

I still don't quite understand the reasoning behind prefix here.

---

*[src/task_behavior_ros/introspection.py, line 156 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAncrKQ0tM_-dzvgGH:-KjAncrKQ0tM_-dzvgGI:bqr9479) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/src/task_behavior_ros/introspection.py#L156)):*
> ```Python
>         self._status_pub.publish(tree_status_msg)
> 
>     def publish_data_dump(self):
> ```

I think you should be passing in the root node you'd like here, default to parent.  That way you can get a dump of the section of the tree you'd like.

---

*[src/task_behavior_ros/introspection.py, line 170 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAo_YOcQKwbPwBg4DT:-KjAo_YOcQKwbPwBg4DU:bu3cy6s) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/src/task_behavior_ros/introspection.py#L170)):*
> ```Python
>         @returns [NodeDataDump] The blackboard data from the node
>         """
>         nodedata = node._blackboard.get_memory(node._id)
> ```

I think it would be worth making a node.get_nodedata() function in task_behavior_engine, instead of calling its 'private' members.

---

*[src/task_behavior_ros/introspection.py, line 200 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAgR-8sdeA99qFXw2U-r1-200:-KjArygqqIlfdiBpy0LT:b5jjje2) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/src/task_behavior_ros/introspection.py#L200)):*
<details><summary><i>Previously, LauraStelzner-TRI wrote…</i></summary><blockquote>

I C&P from the other functions as a starting point, my understanding is it was designed that way because its a recursive function.
</blockquote></details>

Why don't you just call the other recursive function you C&P from or create a common function they both share from?  (or my suggestion in msgs)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10)*

**Comment by LauraStelzner-TRI** [comment]
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[task_behavior_ros.rosinstall, line 11 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAmeGsZ7wXCG33fhxX:-KjB0_rnXVanuEMG4Obu:b-896fix) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/task_behavior_ros.rosinstall#L11)):*
<details><summary><i>Previously, athackst (Allison Thackston) wrote…</i></summary><blockquote>

this will need to be removed (also this one shouldn't be necessary as CI should pull/build the branch you are on)
</blockquote></details>

Done.

---

*[src/task_behavior_ros/introspection.py, line 35 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAnXQk7N5rOPPoJk_G:-KjB0hWDDqhxm7NgZlw_:b-xk1305) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/src/task_behavior_ros/introspection.py#L35)):*
<details><summary><i>Previously, athackst (Allison Thackston) wrote…</i></summary><blockquote>

I still don't quite understand the reasoning behind prefix here.
</blockquote></details>

tree/<topic> is what the tree_viewer expects. If you want to run tree_viewer and have a second introspection object that focuses on a particular branch/sub-branch, then you can set the prefix   

---

*[src/task_behavior_ros/introspection.py, line 156 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAncrKQ0tM_-dzvgGH:-KjBK2oGAYwButjO_QPe:b-896fix) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/src/task_behavior_ros/introspection.py#L156)):*
<details><summary><i>Previously, athackst (Allison Thackston) wrote…</i></summary><blockquote>

I think you should be passing in the root node you'd like here, default to parent.  That way you can get a dump of the section of the tree you'd like.
</blockquote></details>

Done.

---

*[src/task_behavior_ros/introspection.py, line 170 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAo_YOcQKwbPwBg4DT:-KjBK4V5PSRAjQ1K9q3G:b-xrclu2) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/src/task_behavior_ros/introspection.py#L170)):*
<details><summary><i>Previously, athackst (Allison Thackston) wrote…</i></summary><blockquote>

I think it would be worth making a node.get_nodedata() function in task_behavior_engine, instead of calling its 'private' members.
</blockquote></details>

Done.  Will need a new PR for that repo.

---

*[src/task_behavior_ros/introspection.py, line 200 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAgR-8sdeA99qFXw2U-r1-200:-KjBKAW8tH9RBxQAe1_I:b-doomz5) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/src/task_behavior_ros/introspection.py#L200)):*
<details><summary><i>Previously, athackst (Allison Thackston) wrote…</i></summary><blockquote>

Why don't you just call the other recursive function you C&P from or create a common function they both share from?  (or my suggestion in msgs)
</blockquote></details>

I think I've updated it to what you were getting at

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10)*

**Comment by tinawright** [comment]
Reviewed 1 of 4 files at r1, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[src/task_behavior_ros/introspection.py, line 37 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjEYW-YxAS0zLmry-18:-KjEYW-YxAS0zLmry-19:bh8tfli) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L37)):*
> ```Python
>     def __init__(self, node, prefix="tree"):
>         """ recursively set up the pub/subs of this node
>             @param node [Node] The node to recursively introspect
> ```

add @param for prefix, please

---

*[src/task_behavior_ros/introspection.py, line 156 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjEYeEAz2Qzr4CmHuEH:-KjEYeEAz2Qzr4CmHuEI:byuevlm) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L156)):*
> ```Python
>         self._status_pub.publish(tree_status_msg)
> 
>     def publish_data_dump(self, node=None):
> ```

why not just use self.parent here as the default? Then error out below if None is passed in. (I assume parent is not being mutated, hence the difference between this and the ones below.

---

*[src/task_behavior_ros/introspection.py, line 183 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjE_dXTq4f1z-w_tG-q:-KjE_dXUGSKkcleTcLZS:bu37t22) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L183)):*
> ```Python
>         return ndd
> 
>     def _create_tree_data_dump_msg(self, node, msg=TreeDataDump()):
> ```

TreeDataDump is mutable right? you will get the same object for all calls of this method if so.

http://python-guide-pt-br.readthedocs.io/en/latest/writing/gotchas/

---

*[src/task_behavior_ros/introspection.py, line 194 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjE_lrfsZd-4jhsvXbj:-KjE_lrfsZd-4jhsvXbk:b-mdjnve) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L194)):*
> ```Python
>         return msg
> 
>     def _create_status_msg(self, node, msg=TreeStatus()):
> ```

same mutable issue here

---

*[src/task_behavior_ros/introspection.py, line 208 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjE_G-141mRbz4QcEBU:-KjE_G-141mRbz4QcEBV:byhwhnd) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L208)):*
> ```Python
>         msg.name.append(node._name)
>         msg.status.append(node_status_msg)
>         
> ```

FYI excess whitespace

---

*[test/test_introspection.py, line 88 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjEYDKL50EB9UWuY9fw:-KjEYDKL50EB9UWuY9fx:b-kgsflv) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/test/test_introspection.py#L88)):*
> ```Python
> 
>        
>         
> ```

FYI: lots of excess whitespace in this file

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10)*

**Comment by athackst** [comment]
Reviewed 1 of 4 files at r1, 2 of 3 files at r4.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[task_behavior_ros.rosinstall, line 4 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjEaYvR4sQY7xaPxILT:-KjEaYvSQqxjjHUpDXdP:b-l35sj5) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/task_behavior_ros.rosinstall#L4)):*
> ```
>     local-name: task_behavior_engine
>     uri: https://github.com/ToyotaResearchInstitute/task_behavior_engine.git
>     version: node_data_dump
> ```

this will need to be removed

---

*[src/task_behavior_ros/introspection.py, line 35 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAnXQk7N5rOPPoJk_G:-KjEaksu1OC_b4AasYn9:b-akzn4u) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/src/task_behavior_ros/introspection.py#L35)):*
<details><summary><i>Previously, LauraStelzner-TRI wrote…</i></summary><blockquote>

tree/<topic> is what the tree_viewer expects. If you want to run tree_viewer and have a second introspection object that focuses on a particular branch/sub-branch, then you can set the prefix   
</blockquote></details>

Not blocking, but I guess I just don't understand why this would be needed.  Wouldn't it be better to have the visualization tool be able to show your branch/sub-branch and collapse others?

---

*[src/task_behavior_ros/introspection.py, line 122 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjEeDNzQT8XoJDKTQ9w:-KjEeDNzQT8XoJDKTQ9x:b-hjkp4) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L122)):*
> ```Python
>         self._structure_pub.publish(tree_structure_msg)
> 
>     def _create_structure_msg(self, node, msg=TreeStructure()):
> ```

same mutable issue here (oops)

---

*[src/task_behavior_ros/introspection.py, line 177 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjEcMEhP3W582xsM4-R:-KjEcMEhP3W582xsM4-S:blkxbza) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L177)):*
> ```Python
>             ndd.key.append(str(key))
>             ndd.value.append(str(nodedata.get_data(key)))
>         status = node.get_status()
> ```

You can remove this potion, since status already in _create_status_msg

---

*[src/task_behavior_ros/introspection.py, line 183 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjE_dXTq4f1z-w_tG-q:-KjEe5-1wrXnMDM9oa4q:b-4tiinq) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L183)):*
<details><summary><i>Previously, tinawright (Tina Wright (TRI)) wrote…</i></summary><blockquote>

TreeDataDump is mutable right? you will get the same object for all calls of this method if so.

http://python-guide-pt-br.readthedocs.io/en/latest/writing/gotchas/
</blockquote></details>

Good catch.  I just never used the defaults.

---

*[test/test_introspection.py, line 80 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjEedZkuE0LJIvcBQ4g:-KjEedZkuE0LJIvcBQ4h:b4c9lyr) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/test/test_introspection.py#L80)):*
> ```Python
>         self.test._reload()
>         msg = rospy.wait_for_message('tree/structure', TreeStructure)
>         self.check_structure_msg(msg)
> ```

:thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10)*

**Comment by LauraStelzner-TRI** [comment]
Review status: 1 of 4 files reviewed at latest revision, 11 unresolved discussions.

---

*[task_behavior_ros.rosinstall, line 7 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAmaNZg0TeAjl2q6Jw:-KjFF0OM6q5wvnlqNf8s:b-896fix) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/task_behavior_ros.rosinstall#L7)):*
<details><summary><i>Previously, athackst (Allison Thackston) wrote…</i></summary><blockquote>

this will need to be removed
</blockquote></details>

Done.

---

*[task_behavior_ros.rosinstall, line 4 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjEaYvR4sQY7xaPxILT:-KjFF07NatDhPBM3SwQm:b-896fix) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/task_behavior_ros.rosinstall#L4)):*
<details><summary><i>Previously, athackst (Allison Thackston) wrote…</i></summary><blockquote>

this will need to be removed
</blockquote></details>

Done.

---

*[src/task_behavior_ros/introspection.py, line 35 at r1](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjAnXQk7N5rOPPoJk_G:-KjFDJkd7EX23b0mQO-V:b-1zyzi1) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/0c3235fea63710267511b46286ea0672ade40169/src/task_behavior_ros/introspection.py#L35)):*
<details><summary><i>Previously, athackst (Allison Thackston) wrote…</i></summary><blockquote>

Not blocking, but I guess I just don't understand why this would be needed.  Wouldn't it be better to have the visualization tool be able to show your branch/sub-branch and collapse others?
</blockquote></details>

I'm thinking of the case where you have a subscriber within a node, looking at some of the introspection topics, and that node is only interested in a specific branch/sub tree. 
 Because of how these PRs have progressed we don't currently need this option, its more of a potential nice to have.

---

*[src/task_behavior_ros/introspection.py, line 37 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjEYW-YxAS0zLmry-18:-KjFEnyO8ASl0BLc9RuT:b-896fix) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L37)):*
<details><summary><i>Previously, tinawright (Tina Wright (TRI)) wrote…</i></summary><blockquote>

add @param for prefix, please
</blockquote></details>

Done.

---

*[src/task_behavior_ros/introspection.py, line 122 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjEeDNzQT8XoJDKTQ9w:-KjFCZ8zoRCGMAXuTah_:b-896fix) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L122)):*
<details><summary><i>Previously, athackst (Allison Thackston) wrote…</i></summary><blockquote>

same mutable issue here (oops)
</blockquote></details>

Done.

---

*[src/task_behavior_ros/introspection.py, line 156 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjEYeEAz2Qzr4CmHuEH:-KjF8KcKGj-drV3hX8wA:b-jq75zl) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L156)):*
<details><summary><i>Previously, tinawright (Tina Wright (TRI)) wrote…</i></summary><blockquote>

why not just use self.parent here as the default? Then error out below if None is passed in. (I assume parent is not being mutated, hence the difference between this and the ones below.
</blockquote></details>

Tried and it failed the unit test with that change in there     def publish_data_dump(self, node=self.parent):
NameError: name 'self' is not defined

---

*[src/task_behavior_ros/introspection.py, line 177 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjEcMEhP3W582xsM4-R:-KjF90RZVmeLWvHsR920:b-896fix) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L177)):*
<details><summary><i>Previously, athackst (Allison Thackston) wrote…</i></summary><blockquote>

You can remove this potion, since status already in _create_status_msg
</blockquote></details>

Done.

---

*[src/task_behavior_ros/introspection.py, line 183 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjE_dXTq4f1z-w_tG-q:-KjFBI5w6l8ruCLI4mmA:b-896fix) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L183)):*
<details><summary><i>Previously, athackst (Allison Thackston) wrote…</i></summary><blockquote>

Good catch.  I just never used the defaults.
</blockquote></details>

Done.

---

*[src/task_behavior_ros/introspection.py, line 194 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjE_lrfsZd-4jhsvXbj:-KjFEpmvF0bChZ7RcUM5:b-896fix) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L194)):*
<details><summary><i>Previously, tinawright (Tina Wright (TRI)) wrote…</i></summary><blockquote>

same mutable issue here
</blockquote></details>

Done.

---

*[src/task_behavior_ros/introspection.py, line 208 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjE_G-141mRbz4QcEBU:-KjF9UB09tAykIHQL1H0:b-896fix) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L208)):*
<details><summary><i>Previously, tinawright (Tina Wright (TRI)) wrote…</i></summary><blockquote>

FYI excess whitespace
</blockquote></details>

Done.

---

*[test/test_introspection.py, line 88 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjEYDKL50EB9UWuY9fw:-KjF9w6IdQ8rXyBFRNBJ:b-438owv) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/test/test_introspection.py#L88)):*
<details><summary><i>Previously, tinawright (Tina Wright (TRI)) wrote…</i></summary><blockquote>

FYI: lots of excess whitespace in this file
</blockquote></details>

removed where I thought appropriate.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10)*

**Comment by tinawright** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[src/task_behavior_ros/introspection.py, line 156 at r4](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-KjEYeEAz2Qzr4CmHuEH:-KjFoah_qcO9nTYE_O1d:b9odimk) ([raw file](https://github.com/toyotaresearchinstitute/task_behavior_ros/blob/049ff68d99567ac1376ffa259fc620b017730e8f/src/task_behavior_ros/introspection.py#L156)):*
<details><summary><i>Previously, LauraStelzner-TRI wrote…</i></summary><blockquote>

Tried and it failed the unit test with that change in there     def publish_data_dump(self, node=self.parent):
NameError: name 'self' is not defined

</blockquote></details>

good reason.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-:-KjFonPeFbxK1j1ppEpG:bnfp4nl)*

**Comment by athackst** [comment]
:lgtm:

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toyotaresearchinstitute/task_behavior_ros/10#-:-KjJYJ78o1pAu87bTy3J:bnfp4nl)*

**Pull request merged**